File: 17-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.cpp
class Solution {
public:
    // Global
    vector<string> keys = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
    vector<string> ans;

    void solve(string &digits, string &temp, int index){
        int n = digits.size();
        if (index == n){
            ans.push_back(temp);
            return ;
        }


        for(auto &it:keys[digits[index] - '0']){
            temp.push_back(it);
            solve(digits, temp, index + 1);
            temp.pop_back();
        }
    }
    vector<string> letterCombinations(string digits) {
        if (digits.length() == 0) return ans;
        string temp;
        solve(digits, temp, 0);
        return ans;
    }
};

File: 17-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.go
var keys = []string {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"}
func solve(digits string, idx int, comb []byte, poss_comb *[]string) {
    if idx == len(digits) {
        s := string(comb)
        *poss_comb = append(*poss_comb, s)
        return 
    }

    key := keys[(digits[idx] - '0')]
    for i := range key {
        comb = append(comb, key[i])
        solve(digits, idx+1, comb, poss_comb)
        comb = comb[:len(comb)-1]
    }

}

func letterCombinations(digits string) []string {
    if len(digits) == 0 {
        return []string{}
    }
    var ans []string
    var temp []byte
    solve(digits, 0, temp, &ans)
    return ans
}